module/zfs/dsl_synctask.c

```diff
@@ -143,9 +146,12 @@ void
 dsl_sync_task_sync(dsl_sync_task_t *dst, dmu_tx_t *tx)
 {
 	dsl_pool_t *dp = dst->dst_pool;
+	metaslab_class_t *mc;
 
 	ASSERT0(dst->dst_error);
 
+	mc = spa_preferred_class(dp->dp_spa, dst->dst_space, DMU_OT_NONE, 0);
```

Comment:
I think you actually always want the normal class here.  Or maybe you want to add up the deferred space of all the classes?  In any case, it shouldn't depend on dst_space.